### ISSUE TITLE
FreeOrion app hangs on program exit on OSX

### ISSUE THREAD

**Issue by Vezzra** [freeorion/freeorion#564]
Commit 545449ccb53cf3bf88f17e05c6d5b50cdec1e0c2 (part of PR #519) introduced a weird bug on OSX: when exiting the application, it does not terminate cleanly, but hangs. The only way to terminate the program completely is to kill the process.

When starting the program from the command line, I get the following output in the shell immediately after program launch:

```
2016-03-25 13:27:04.712 freeorion[13473:87446] _createMenuRef called with existing principal MenuRef already associated with menu
2016-03-25 13:27:04.713 freeorion[13473:87446] (
    0   CoreFoundation                      0x96686c63 __raiseError + 195
    1   libobjc.A.dylib                     0x932e9a2a objc_exception_throw + 276
    2   CoreFoundation                      0x96686b7d +[NSException raise:format:] + 141
    3   AppKit                              0x9ad2a5b3 -[NSCarbonMenuImpl _createMenuRef] + 70
    4   AppKit                              0x9ad29d44 -[NSCarbonMenuImpl _instantiateCarbonMenu] + 155
    5   AppKit                              0x9ad29c9e -[NSCarbonMenuImpl setupCarbonMenuBar] + 40
    6   AppKit                              0x9ad281fd -[NSApplication finishLaunching] + 1398
    7   AppKit                              0x9ad278a0 -[NSApplication run] + 143
    8   AppKit                              0x9ac9cfa0 NSApplicationMain + 2082
    9   freeorion                           0x00079f96 main + 408
    10  freeorion                           0x0003e385 start + 53
)
```

freeorion.log contents:

```
2016-03-25 13:27:05.465621 [debug] Client : Logger initialized
2016-03-25 13:27:05.465786 [debug] Client : v0.4.5+ HEAD [build 2016-03-12.545449c] Xcode 0640
2016-03-25 13:27:05.465828 [debug] Client : GL Version String: 2.1 ATI-1.32.25
2016-03-25 13:27:05.465878 [debug] Client : Dependency versions from headers:
2016-03-25 13:27:05.465890 [debug] Client : Boost: 1_56
2016-03-25 13:27:05.465901 [debug] Client : FreeType: 2.3.11
2016-03-25 13:27:05.465912 [debug] Client : PNG: 1.5.12
2016-03-25 13:27:05.465922 [debug] Client : Python: 2.7.6
2016-03-25 13:27:05.465933 [debug] Client : SDL: 2.0.3
2016-03-25 13:27:05.465944 [debug] Client : zlib: 1.2.7
2016-03-25 13:27:05.665122 [debug] Client : OpenAL initialized. Version 1.1Renderer SoftwareVendor Apple Computer Inc.
Extensions: AL_EXT_OFFSET AL_EXT_LINEAR_DISTANCE AL_EXT_EXPONENT_DISTANCE AL_EXT_float32 AL_EXT_STATIC_BUFFER AL_EXT_SOURCE_NOTIFICATIONS AL_EXT_SOURCE_SPATIALIZATION

2016-03-25 13:27:06.123189 [debug] Client : Tech Tree Layout Preparing Tech Data
2016-03-25 13:27:07.923312 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:07.932034 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.505051 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.507530 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.509998 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.516545 [debug] Client : Tech Tree Layout Doing Graph Layout
2016-03-25 13:27:08.516576 [debug] Client : TechTreeLayout::DoLayout creaing place holder nodes...
2016-03-25 13:27:08.518655 [debug] Client : Tech Tree Layout Creating Panels
2016-03-25 13:27:08.520369 [debug] Client : Tech Tree Layout Done
2016-03-25 13:27:08.520418 [debug] Client : Tech List Box Populating
2016-03-25 13:27:08.577871 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.583095 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.683817 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.684750 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.685631 [debug] Client : SourceForEmpire: Unable to get empire with ID: -1
2016-03-25 13:27:08.710990 [debug] Client : Tech List Box Done Populating
2016-03-25 13:27:08.711016 [debug] Client :     Creation time=165.985ms
2016-03-25 13:27:08.711036 [debug] Client :     Insertion time=20.808ms
2016-03-25 13:27:08.839665 [debug] Client : BasesListBox::Populate
2016-03-25 13:27:08.839709 [debug] Client : BasesListBox::PopulateWithSavedDesigns
2016-03-25 13:27:08.841323 [debug] Client : DesignWnd::MainPanel::ReregisterDesigns
2016-03-25 13:27:08.985701 [debug] Client : Limited FPS to 60
2016-03-25 13:27:08.986056 [debug] Client : OpenGL Version Number: 2.1
2016-03-25 13:27:08.986497 [debug] Client : (HumanClientFSM) IntroMenu
2016-03-25 13:27:08.986520 [debug] Client : HumanClientApp::KillServer()
2016-03-25 13:27:08.986535 [debug] Client : Process::Kill
2016-03-25 13:27:08.986547 [debug] Client : Process::Kill found no m_impl
2016-03-25 13:27:08.986557 [debug] Client : Process::Kill calling RequestTermination()
2016-03-25 13:27:08.986570 [debug] Client : ClientNetworking::SetPlayerID: player id set to: -1
2016-03-25 13:27:09.014411 [debug] Client : HumanClientApp::HandleFocusChange(Gained Focus)
2016-03-25 13:27:10.054480 [debug] Client : HumanClientApp::HandleFocusChange(Lost Focus)
2016-03-25 13:27:13.110366 [debug] Client : HumanClientApp::HandleFocusChange(Gained Focus)
2016-03-25 13:27:14.973401 [debug] Client : (HumanClientFSM) ~IntroMenu
```

Build used: 545449ccb53cf3bf88f17e05c6d5b50cdec1e0c2

I'm a bit stumped because I can't imagine how the changes of that commit could have caused this. @LGM-Doyle, any flashes of ingenious insights...?

**Comment by LGM-Doyle** [comment]
Vezzra,

Please read and try the next comment before this one.

I have created PR #568 in an attempt to fix this.  I can't test it as I am running linux.  Could you test it and provide me feedback.

Right now, this is what I think the problem is.

Windows, linux and OSX do static initialization prior to main() of the graphics system differently.  SDL expects that all applications substitute main() with SDL_main().  Freeorion has never done that, and it was not a problem.  I added some static initialization of the SDL video subsystem in order to detect the screen size before the app was started up.  Static initialization timings are not specified in C++.  I think that OSX is initializing part of the SDL system in the HumanClient static initialization code before its own static initialization.

I've moved the initialization into main.  I hope this fixes the problem.

**Comment by LGM-Doyle** [comment]
Vezzra,
Could you revert just grooming 3ce77ed9b9222567ee357bfbab4efe7722847f85 from a few days ago and see if that alone fixes the crash.
The Exit(0) on line 855 prevents the app from exiting, but stops the screen from updating.

**Comment by Vezzra** [comment]
> Please read and try the next comment before this one.

Oops, sorry, saw the PR first, and tested and merged it before I saw your comments here. Will test your second solution too.

**Comment by Vezzra** [comment]
> Could you revert just grooming 3ce77ed from a few days ago and see if that alone fixes the crash.
> The Exit(0) on line 855 prevents the app from exiting, but stops the screen from updating.

Nope, reverting 3ce77ed doesn't fix the crash (on the contrary, with #568 applied, everything works fine even with the changes of 3ce77ed).

Just curious: why do you think it would? 3ce77ed didn't introduce the issue...

**Comment by Vezzra** [comment]
Fixed by #568.

**Comment by LGM-Doyle** [comment]
My bad. 

I have submitted a second patch https://github.com/freeorion/freeorion/pull/570 to fix the second independent but similar problem.

Here is the story.  

I trusted your assessment of which PR caused the crash on OSX and wrote the patch. I couldn't test it as I had no crash on my machine. and pushed it. While pushing it, I fetched and merged with master.

After I pushed it I noticed that 3ce77ed caused a similar crash on exit on my linux machine.  I no longer trusted that my speculative patch fixed anything.  I also no longer completely trusted which patch caused the crash because I had evidence of a similar crash from the innocuously named grooming patch on the same file that caused your reported issue.  

I thought I had pushed a pointless patch.

In the end, two problems, two patches.
